File: src/lws/cli/services/stepfunctions.py
"""``lws stepfunctions`` sub-commands."""

from __future__ import annotations

import asyncio

import typer

from lws.cli.services.client import (
    LwsClient,
    exit_with_error,
    json_request_output,
    output_json,
    parse_json_option,
)

app = typer.Typer(help="Step Functions commands")

_SERVICE = "stepfunctions"
_TARGET_PREFIX = "AWSStepFunctions"


def _client(port: int) -> LwsClient:
    return LwsClient(port=port)


@app.command("start-execution")
def start_execution(
    name: str = typer.Option(..., "--name", help="State machine name"),
    execution_input: str = typer.Option("{}", "--input", help="JSON input"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Start a state machine execution."""
    asyncio.run(_start_execution(name, execution_input, port))


async def _start_execution(name: str, input_str: str, port: int) -> None:
    client = _client(port)
    try:
        resource = await client.resolve_resource(_SERVICE, name)
        arn = resource.get("arn", f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}")
    except Exception:
        arn = f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}"
    result = await client.json_target_request(
        _SERVICE,
        f"{_TARGET_PREFIX}.StartExecution",
        {"stateMachineArn": arn, "input": input_str},
    )
    output_json(result)


@app.command("describe-execution")
def describe_execution(
    execution_arn: str = typer.Option(..., "--execution-arn", help="Execution ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Describe a state machine execution."""
    asyncio.run(_describe_execution(execution_arn, port))


async def _describe_execution(execution_arn: str, port: int) -> None:
    client = _client(port)
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.DescribeExecution",
            {"executionArn": execution_arn},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("list-executions")
def list_executions(
    name: str = typer.Option(..., "--name", help="State machine name"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List executions for a state machine."""
    asyncio.run(_list_executions(name, port))


async def _list_executions(name: str, port: int) -> None:
    client = _client(port)
    try:
        resource = await client.resolve_resource(_SERVICE, name)
        arn = resource.get("arn", f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}")
    except Exception:
        arn = f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}"
    result = await client.json_target_request(
        _SERVICE,
        f"{_TARGET_PREFIX}.ListExecutions",
        {"stateMachineArn": arn},
    )
    output_json(result)


@app.command("list-state-machines")
def list_state_machines(
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List all state machines."""
    asyncio.run(_list_state_machines(port))


async def _list_state_machines(port: int) -> None:
    client = _client(port)
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.ListStateMachines",
            {},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("create-state-machine")
def create_state_machine(
    name: str = typer.Option(..., "--name", help="State machine name"),
    definition: str = typer.Option(..., "--definition", help="ASL definition JSON"),
    role_arn: str = typer.Option("", "--role-arn", help="IAM role ARN"),
    sm_type: str = typer.Option("STANDARD", "--type", help="STANDARD or EXPRESS"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Create a state machine."""
    asyncio.run(_create_state_machine(name, definition, role_arn, sm_type, port))


async def _create_state_machine(
    name: str, definition: str, role_arn: str, sm_type: str, port: int
) -> None:
    client = _client(port)
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.CreateStateMachine",
            {
                "name": name,
                "definition": definition,
                "roleArn": role_arn,
                "type": sm_type,
            },
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("delete-state-machine")
def delete_state_machine(
    name: str = typer.Option(..., "--name", help="State machine name"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Delete a state machine."""
    asyncio.run(_delete_state_machine(name, port))


async def _delete_state_machine(name: str, port: int) -> None:
    client = _client(port)
    arn = f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}"
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.DeleteStateMachine",
            {"stateMachineArn": arn},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("describe-state-machine")
def describe_state_machine(
    name: str = typer.Option(..., "--name", help="State machine name"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Describe a state machine."""
    asyncio.run(_describe_state_machine(name, port))


async def _describe_state_machine(name: str, port: int) -> None:
    client = _client(port)
    arn = f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}"
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.DescribeStateMachine",
            {"stateMachineArn": arn},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("start-sync-execution")
def start_sync_execution(
    name: str = typer.Option(..., "--name", help="State machine name"),
    execution_input: str = typer.Option("{}", "--input", help="JSON input"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Start a synchronous state machine execution."""
    asyncio.run(_start_sync_execution(name, execution_input, port))


async def _start_sync_execution(name: str, input_str: str, port: int) -> None:
    client = _client(port)
    try:
        resource = await client.resolve_resource(_SERVICE, name)
        arn = resource.get("arn", f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}")
    except Exception:
        arn = f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}"
    result = await client.json_target_request(
        _SERVICE,
        f"{_TARGET_PREFIX}.StartSyncExecution",
        {"stateMachineArn": arn, "input": input_str},
    )
    output_json(result)


@app.command("stop-execution")
def stop_execution(
    execution_arn: str = typer.Option(..., "--execution-arn", help="Execution ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Stop a state machine execution."""
    asyncio.run(_stop_execution(execution_arn, port))


async def _stop_execution(execution_arn: str, port: int) -> None:
    client = _client(port)
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.StopExecution",
            {"executionArn": execution_arn},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("update-state-machine")
def update_state_machine(
    name: str = typer.Option(..., "--name", help="State machine name"),
    definition: str = typer.Option(..., "--definition", help="ASL definition JSON"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Update a state machine."""
    asyncio.run(_update_state_machine(name, definition, port))


async def _update_state_machine(name: str, definition: str, port: int) -> None:
    client = _client(port)
    arn = f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}"
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.UpdateStateMachine",
            {"stateMachineArn": arn, "definition": definition},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("get-execution-history")
def get_execution_history(
    execution_arn: str = typer.Option(..., "--execution-arn", help="Execution ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Get execution history."""
    asyncio.run(_get_execution_history(execution_arn, port))


async def _get_execution_history(execution_arn: str, port: int) -> None:
    client = _client(port)
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.GetExecutionHistory",
            {"executionArn": execution_arn},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("validate-state-machine-definition")
def validate_state_machine_definition(
    definition: str = typer.Option(..., "--definition", help="ASL definition JSON"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Validate a state machine definition."""
    asyncio.run(_validate_state_machine_definition(definition, port))


async def _validate_state_machine_definition(definition: str, port: int) -> None:
    client = _client(port)
    try:
        result = await client.json_target_request(
            _SERVICE,
            f"{_TARGET_PREFIX}.ValidateStateMachineDefinition",
            {"definition": definition},
        )
    except Exception as exc:
        exit_with_error(str(exc))
    output_json(result)


@app.command("list-state-machine-versions")
def list_state_machine_versions(
    name: str = typer.Option(..., "--name", help="State machine name"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List state machine versions."""
    asyncio.run(_list_state_machine_versions(name, port))


async def _list_state_machine_versions(name: str, port: int) -> None:
    arn = f"arn:aws:states:us-east-1:000000000000:stateMachine:{name}"
    await json_request_output(
        port,
        _SERVICE,
        f"{_TARGET_PREFIX}.ListStateMachineVersions",
        {"stateMachineArn": arn},
    )


@app.command("tag-resource")
def tag_resource(
    resource_arn: str = typer.Option(..., "--resource-arn", help="State machine or activity ARN"),
    tags: str = typer.Option(..., "--tags", help="JSON array of key/value tag objects"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Tag a Step Functions resource."""
    parsed_tags = parse_json_option(tags, "--tags")
    asyncio.run(
        json_request_output(
            port,
            _SERVICE,
            f"{_TARGET_PREFIX}.TagResource",
            {"resourceArn": resource_arn, "tags": parsed_tags},
        )
    )


@app.command("untag-resource")
def untag_resource(
    resource_arn: str = typer.Option(..., "--resource-arn", help="State machine or activity ARN"),
    tag_keys: str = typer.Option(..., "--tag-keys", help="JSON array of tag key strings to remove"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """Untag a Step Functions resource."""
    parsed_keys = parse_json_option(tag_keys, "--tag-keys")
    asyncio.run(
        json_request_output(
            port,
            _SERVICE,
            f"{_TARGET_PREFIX}.UntagResource",
            {"resourceArn": resource_arn, "tagKeys": parsed_keys},
        )
    )


@app.command("list-tags-for-resource")
def list_tags_for_resource(
    resource_arn: str = typer.Option(..., "--resource-arn", help="State machine or activity ARN"),
    port: int = typer.Option(3000, "--port", "-p", help="LDK port"),
) -> None:
    """List tags for a Step Functions resource."""
    asyncio.run(
        json_request_output(
            port,
            _SERVICE,
            f"{_TARGET_PREFIX}.ListTagsForResource",
            {"resourceArn": resource_arn},
        )
    )

File: lang/typescript/core/tests/steps/stepfunctions.ts
/** Step Functions step definitions. */

import { Given, When, Then } from "@cucumber/cucumber";
import assert from "assert";
import {
  CreateStateMachineCommand,
  DeleteStateMachineCommand,
  DescribeStateMachineCommand,
  ListStateMachinesCommand,
  StartExecutionCommand,
  StartSyncExecutionCommand,
  StopExecutionCommand,
  DescribeExecutionCommand,
  GetExecutionHistoryCommand,
  ListExecutionsCommand,
  ListStateMachineVersionsCommand,
  TagResourceCommand,
  UntagResourceCommand,
  ListTagsForResourceCommand,
  UpdateStateMachineCommand,
  ValidateStateMachineDefinitionCommand,
} from "@aws-sdk/client-sfn";
import type { LwsWorld } from "../support/world";

const ACCOUNT = "000000000000";
const REGION = "us-east-1";

const PASS_DEFINITION = JSON.stringify({
  Comment: "A simple pass state machine",
  StartAt: "Pass",
  States: { Pass: { Type: "Pass", End: true } },
});

function smArn(smName: string): string {
  return `arn:aws:states:${REGION}:${ACCOUNT}:stateMachine:${smName}`;
}

async function createStateMachine(world: LwsWorld, smName: string, type = "STANDARD"): Promise<string> {
  const client = world.sfnClient();
  const result = await client.send(
    new CreateStateMachineCommand({
      name: smName,
      definition: PASS_DEFINITION,
      roleArn: `arn:aws:iam::${ACCOUNT}:role/dummy-role`,
      type: type as "STANDARD" | "EXPRESS",
    })
  );
  return result.stateMachineArn!;
}

// --- Given -----------------------------------------------------------------

Given("a state machine {string} was created with a Pass definition", async function (
  this: LwsWorld,
  smName: string
) {
  const arn = await createStateMachine(this, smName);
  this.lastStateMachineArn = arn;
});

Given("an EXPRESS state machine {string} was created with a Pass definition", async function (
  this: LwsWorld,
  smName: string
) {
  const arn = await createStateMachine(this, smName, "EXPRESS");
  this.lastStateMachineArn = arn;
});

Given("an execution was started on state machine {string} with input {string}", async function (
  this: LwsWorld,
  smName: string,
  input: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  const result = await client.send(
    new StartExecutionCommand({ stateMachineArn: arn, input })
  );
  this.lastExecutionArn = result.executionArn;
});

// --- When ------------------------------------------------------------------

When("I create a state machine named {string} with a Pass definition", async function (
  this: LwsWorld,
  smName: string
) {
  const client = this.sfnClient();
  try {
    const result = await client.send(
      new CreateStateMachineCommand({
        name: smName,
        definition: PASS_DEFINITION,
        roleArn: `arn:aws:iam::${ACCOUNT}:role/dummy-role`,
        type: "STANDARD",
      })
    );
    this.lastStateMachineArn = result.stateMachineArn;
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I delete state machine {string}", async function (this: LwsWorld, smName: string) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(new DeleteStateMachineCommand({ stateMachineArn: arn }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I describe state machine {string}", async function (this: LwsWorld, smName: string) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(new DescribeStateMachineCommand({ stateMachineArn: arn }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list state machines", async function (this: LwsWorld) {
  const client = this.sfnClient();
  try {
    const result = await client.send(new ListStateMachinesCommand({}));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I start an execution on state machine {string} with input {string}", async function (
  this: LwsWorld,
  smName: string,
  input: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(new StartExecutionCommand({ stateMachineArn: arn, input }));
    this.lastExecutionArn = result.executionArn;
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I start a sync execution on state machine {string} with input {string}", async function (
  this: LwsWorld,
  smName: string,
  input: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(
      new StartSyncExecutionCommand({ stateMachineArn: arn, input })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I stop the started execution", async function (this: LwsWorld) {
  const client = this.sfnClient();
  try {
    const result = await client.send(
      new StopExecutionCommand({ executionArn: this.lastExecutionArn! })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I describe the started execution", async function (this: LwsWorld) {
  const client = this.sfnClient();
  try {
    const result = await client.send(
      new DescribeExecutionCommand({ executionArn: this.lastExecutionArn! })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I get execution history for the started execution", async function (this: LwsWorld) {
  const client = this.sfnClient();
  try {
    const result = await client.send(
      new GetExecutionHistoryCommand({ executionArn: this.lastExecutionArn! })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list executions for state machine {string}", async function (
  this: LwsWorld,
  smName: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(new ListExecutionsCommand({ stateMachineArn: arn }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list state machine versions for {string}", async function (
  this: LwsWorld,
  smName: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(
      new ListStateMachineVersionsCommand({ stateMachineArn: arn })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I list tags for state machine {string}", async function (
  this: LwsWorld,
  smName: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(new ListTagsForResourceCommand({ resourceArn: arn }));
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I tag state machine {string} with tags [\\{{string}: {string}, {string}: {string}}]", async function (
  this: LwsWorld,
  smName: string,
  _k1: string,
  _v1: string,
  _k2: string,
  _v2: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(
      new TagResourceCommand({ resourceArn: arn, tags: [{ key: "env", value: "test" }] })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

Given("state machine {string} was tagged with tags [\\{{string}: {string}, {string}: {string}}]", async function (
  this: LwsWorld,
  smName: string,
  _k1: string,
  _v1: string,
  _k2: string,
  _v2: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  await client.send(
    new TagResourceCommand({ resourceArn: arn, tags: [{ key: "env", value: "test" }] })
  );
});

When("I untag state machine {string} with tag keys [{string}]", async function (
  this: LwsWorld,
  smName: string,
  _tagKeysStr: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(
      new UntagResourceCommand({ resourceArn: arn, tagKeys: ["env"] })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I update state machine {string} with an updated definition", async function (
  this: LwsWorld,
  smName: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  const updatedDefinition = JSON.stringify({
    Comment: "Updated pass state machine",
    StartAt: "Pass",
    States: { Pass: { Type: "Pass", End: true } },
  });
  try {
    const result = await client.send(
      new UpdateStateMachineCommand({ stateMachineArn: arn, definition: updatedDefinition })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I validate a Pass state machine definition", async function (this: LwsWorld) {
  const client = this.sfnClient();
  try {
    const result = await client.send(
      new ValidateStateMachineDefinitionCommand({ definition: PASS_DEFINITION })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

When("I tag state machine {string} with tags [{string}]", async function (
  this: LwsWorld,
  smName: string,
  _tagsStr: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  try {
    const result = await client.send(
      new TagResourceCommand({ resourceArn: arn, tags: [{ key: "env", value: "test" }] })
    );
    this.lastResult = { success: true, output: result };
  } catch (err) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

// Timed variant
When("I list Step Functions state machines with timing", async function (this: LwsWorld) {
  const client = this.sfnClient();
  const start = Date.now();
  try {
    const result = await client.send(new ListStateMachinesCommand({}));
    this.timedResult = { success: true, output: result, elapsedMs: Date.now() - start };
  } catch (err) {
    this.timedResult = { success: false, output: err, elapsedMs: Date.now() - start };
  }
});

When("I list Step Functions state machines", async function (this: LwsWorld) {
  const client = this.sfnClient();
  try {
    const result = await client.send(new ListStateMachinesCommand({}));
    this.lastResult = { success: true, output: result };
  } catch (err: unknown) {
    this.lastResult = { success: false, output: err, error: err };
  }
});

// --- Then ------------------------------------------------------------------

Then("state machine {string} will exist", async function (this: LwsWorld, smName: string) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  const result = await client.send(new DescribeStateMachineCommand({ stateMachineArn: arn }));
  assert.ok(result.name, "Expected state machine to exist");
});

Then("state machine {string} will not appear in list-state-machines", async function (
  this: LwsWorld,
  smName: string
) {
  const client = this.sfnClient();
  const result = await client.send(new ListStateMachinesCommand({}));
  const names = (result.stateMachines ?? []).map((sm) => sm.name ?? "");
  assert.ok(!names.includes(smName), `Expected state machine "${smName}" to not be in list`);
});

Then("the state machine list will include {string}", function (this: LwsWorld, smName: string) {
  const output = this.lastResult.output as { stateMachines?: Array<{ name?: string }> };
  const names = (output?.stateMachines ?? []).map((sm) => sm.name ?? "");
  assert.ok(names.includes(smName), `Expected state machine list to include "${smName}"`);
});

Then("the output will contain state machine name {string}", function (
  this: LwsWorld,
  smName: string
) {
  const actualOutput = JSON.stringify(this.lastResult.output);
  assert.ok(actualOutput.includes(smName), `Expected output to contain "${smName}"`);
});

Then("the output will contain an execution ARN", function (this: LwsWorld) {
  const output = this.lastResult.output as { executionArn?: string };
  assert.ok(output?.executionArn, "Expected output to contain an executionArn");
});

Then("the output will contain the execution ARN", function (this: LwsWorld) {
  const output = this.lastResult.output as { executionArn?: string };
  assert.ok(output?.executionArn, "Expected output to contain the executionArn");
});

Then("the output will contain a status field", function (this: LwsWorld) {
  const output = this.lastResult.output as { status?: string };
  assert.ok(output?.status, "Expected output to contain a status field");
});

Then("the started execution will have a status", async function (this: LwsWorld) {
  const client = this.sfnClient();
  const result = await client.send(
    new DescribeExecutionCommand({ executionArn: this.lastExecutionArn! })
  );
  assert.ok(result.status, "Expected execution to have a status");
});

Then("the stopped execution will have status {string}", async function (
  this: LwsWorld,
  expectedStatus: string
) {
  const client = this.sfnClient();
  const result = await client.send(
    new DescribeExecutionCommand({ executionArn: this.lastExecutionArn! })
  );
  assert.strictEqual(result.status, expectedStatus);
});

Then("the executions list will have at least {int} entry", function (
  this: LwsWorld,
  minCount: number
) {
  const output = this.lastResult.output as { executions?: unknown[] };
  const count = (output?.executions ?? []).length;
  assert.ok(count >= minCount, `Expected at least ${minCount} executions but got ${count}`);
});

Then("state machine {string} will have tag {string} with value {string}", async function (
  this: LwsWorld,
  smName: string,
  tagKey: string,
  expectedValue: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  const result = await client.send(new ListTagsForResourceCommand({ resourceArn: arn }));
  const tags = result.tags ?? [];
  const tag = tags.find((t) => t.key === tagKey);
  assert.strictEqual(tag?.value, expectedValue);
});

Then("state machine {string} will not have tag {string}", async function (
  this: LwsWorld,
  smName: string,
  tagKey: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  const result = await client.send(new ListTagsForResourceCommand({ resourceArn: arn }));
  const tags = result.tags ?? [];
  const tag = tags.find((t) => t.key === tagKey);
  assert.ok(!tag, `Expected tag "${tagKey}" to not exist`);
});

Then("state machine {string} will have the updated definition", async function (
  this: LwsWorld,
  smName: string
) {
  const client = this.sfnClient();
  const arn = smArn(smName);
  const result = await client.send(new DescribeStateMachineCommand({ stateMachineArn: arn }));
  assert.ok(result.definition, "Expected state machine to have a definition");
});
